libcudacxx/include/cuda/std/__random/philox_engine.h

```diff
@@ -0,0 +1,572 @@
+    if constexpr (word_count == 2)
+    {
+      ::cuda::std::array<result_type, word_count> __S     = __x_;
+      ::cuda::std::array<result_type, word_count / 2> __K = __k_;
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (::cuda::std::size_t __j = 0; __j < round_count; ++__j)
+      {
+        auto [__hi, __lo] = __mulhilo(__S[0], multipliers[0]);
+        __S[0]            = __hi ^ __K[0] ^ __S[1];
+        __S[1]            = __lo;
+        __K[0]            = (__K[0] + round_consts[0]) & max();
+      }
+      __y_ = __S;
+    }
+    else if constexpr (word_count == 4)
+    {
+      ::cuda::std::array<result_type, word_count> __S     = __x_;
+      ::cuda::std::array<result_type, word_count / 2> __K = __k_;
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (::cuda::std::size_t __j = 0; __j < round_count; ++__j)
+      {
+        ::cuda::std::array<result_type, word_count> __V = {__S[2], __S[1], __S[0], __S[3]};
+        auto [__hi0, __lo0]                             = __mulhilo(__V[0], multipliers[1]);
+        auto [__hi2, __lo2]                             = __mulhilo(__V[2], multipliers[0]);
+
+        __S[0] = __hi0 ^ __K[0] ^ __V[1];
+        __S[1] = __lo0;
+
+        __S[2] = __hi2 ^ __K[1] ^ __V[3];
+
+        __S[3] = __lo2;
+
+        __K[0] = (__K[0] + round_consts[0]) & max();
+        __K[1] = (__K[1] + round_consts[1]) & max();
+      }
+
+      __y_ = __S;
+    }
+  }
+
+  // The counter X, a big integer stored as word_count w-bit words.
+  // The least significant word is __x_[0].
+  ::cuda::std::array<result_type, word_count> __x_ = {};
+  // K is the "Key", storing the seed
+  ::cuda::std::array<result_type, word_count / 2> __k_ = {};
+  // The output buffer Y
+  // Each time __j_ reaches word_count, we generate word_count new values and store them in __y_.
+  ::cuda::std::array<result_type, word_count> __y_ = {};
+  // Each generation produces n random numbers, which are returned one at a time.
+  // __j_ cycles through [0, n-1].
+  unsigned long long __j_ = 0;
+
+}; // end philox_engine
+
+//! @typedef philox4x32
+//! @brief A random number engine with predefined parameters which implements the
+//!        Philox counter based random number generation algorithm.
+//! @note The 10000th consecutive invocation of a default-constructed object of type philox4x32
+//!       shall produce the value 1955073260.
+using philox4x32 = philox_engine<std::uint_fast32_t, 32, 4, 10, 0xD2511F53, 0x9E3779B9, 0xCD9E8D57, 0xBB67AE85>;
+
+//! @typedef philox4x64
+//! @brief A random number engine with predefined parameters which implements the
+//!        Philox counter based random number generation algorithm.
+//! @note The 10000th consecutive invocation of a default-constructed object of type philox4x64
+//!       shall produce the value 3409172418970261260.
+using philox4x64 =
+  philox_engine<std::uint_fast64_t, 64, 4, 10, 0xD2E7470EE14C6C93, 0x9E3779B97F4A7C15, 0xCA5A826395121157, 0xBB67AE8584CAA73B>;
```

Comment:
I believe those should be
```suggestion
  philox_engine<::cuda::std::uint_fast64_t, 64, 4, 10, 0xD2E7470EE14C6C93, 0x9E3779B97F4A7C15, 0xCA5A826395121157, 0xBB67AE8584CAA73B>;
```